zenith_utils/src/pq_proto.rs

```diff
@@ -0,0 +1,559 @@
+                io::ErrorKind::InvalidInput,
+                "named prepared statements not implemented in Describe",
+            ));
+        }
+        */
+
+        if kind != b'S' {
+            bail!("only prepared statmement Describe is implemented");
+        }
+
+        Ok(FeMessage::Describe(FeDescribeMessage { kind }))
+    }
+}
+
+impl FeExecuteMessage {
+    pub fn parse(mut buf: Bytes) -> anyhow::Result<FeMessage> {
+        let portal_name = read_null_terminated(&mut buf)?;
+        let maxrows = buf.get_i32();
+
+        if !portal_name.is_empty() {
+            bail!("named portals not implemented");
+        }
+
+        if maxrows != 0 {
+            bail!("row limit in Execute message not supported");
+        }
+
+        Ok(FeMessage::Execute(FeExecuteMessage { maxrows }))
+    }
+}
+
+impl FeBindMessage {
+    pub fn parse(mut buf: Bytes) -> anyhow::Result<FeMessage> {
+        let portal_name = read_null_terminated(&mut buf)?;
+        let _pstmt_name = read_null_terminated(&mut buf)?;
+
+        if !portal_name.is_empty() {
+            bail!("named portals not implemented");
+        }
+
+        // FIXME: see FeParseMessage::parse
+        /*
+        if !pstmt_name.is_empty() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "named prepared statements not implemented",
+            ));
+        }
+        */
+
+        Ok(FeMessage::Bind(FeBindMessage {}))
+    }
+}
+
+impl FeCloseMessage {
+    pub fn parse(mut buf: Bytes) -> anyhow::Result<FeMessage> {
+        let _kind = buf.get_u8();
+        let _pstmt_or_portal_name = read_null_terminated(&mut buf)?;
+
+        // FIXME: we do nothing with Close
+
+        Ok(FeMessage::Close(FeCloseMessage {}))
+    }
+}
+
+fn read_null_terminated(buf: &mut Bytes) -> anyhow::Result<Bytes> {
+    let mut result = BytesMut::new();
+
+    loop {
+        if !buf.has_remaining() {
+            bail!("no null-terminator in string");
+        }
+
+        let byte = buf.get_u8();
+
+        if byte == 0 {
+            break;
+        }
+        result.put_u8(byte);
+    }
+    Ok(result.freeze())
+}
+
+#[derive(Debug)]
+pub enum BeMessage<'a> {
+    AuthenticationOk,
+    BindComplete,
+    CommandComplete(&'a [u8]),
+    ControlFile,
+    CopyData(&'a [u8]),
+    CopyDone,
+    CopyOutResponse,
+    CopyBothResponse,
+    CloseComplete,
+    // None means column is NULL
+    DataRow(&'a [Option<&'a [u8]>]),
+    ErrorResponse(String),
+    // see https://www.postgresql.org/docs/devel/protocol-flow.html#id-1.10.5.7.11
+    Negotiate,
+    NoData,
+    ParameterDescription,
+    ParameterStatus,
+    ParseComplete,
+    ReadyForQuery,
+    RowDescription(&'a [RowDescriptor]),
+    XLogData(XLogDataBody<'a>),
+}
+
+#[derive(Debug)]
+pub struct RowDescriptor {
+    pub name: &'static [u8],
+    pub tableoid: Oid,
+    pub attnum: i16,
+    pub typoid: Oid,
+    pub typlen: i16,
+    pub typmod: i32,
+    pub formatcode: i16,
```

Comment:
Should we add comments for the meanings of these

Reply:
They are from the protocol, added comment about that.